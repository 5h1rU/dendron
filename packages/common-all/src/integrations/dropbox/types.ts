import { files } from "dropbox";

export type MetadataReference =
  | files.FileMetadataReference
  | files.FolderMetadataReference;

export interface ListFolderResultSimple {
  /**
   * The files and (direct) subfolders in the folder.
   */
  entries: Array<MetadataReference>;
  /**
   * Pass the cursor into listFolderContinue() to see what's changed in the
   * folder since your previous query.
   */
  cursor: files.ListFolderCursor;
  /**
   * If true, then there are more entries available. Pass the cursor to
   * listFolderContinue() to retrieve the rest.
   */
  has_more: boolean;
}
