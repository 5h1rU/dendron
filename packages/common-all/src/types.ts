// Global Types

import { Note, Schema } from "./node";

export type Stage = "dev" | "prod" | "test";

// === Node Types

// --- Primitives
export type NoteData = {
  schemaId?: string;
};
export type SchemaData = {
  namespace?: boolean;
};
export type DNodeData = SchemaData | NoteData;
export type IDNodeType = "note" | "schema";

export type QueryMode = IDNodeType;

// --- Nodes
export type DNodeRawOpts<T extends DNodeData> = {
  id?: string;
  title?: string;
  desc?: string;
  stub?: boolean;
  updated?: string;
  created?: string;
  /**
   * Same as uri
   */
  fname?: string;
  parent?: string | null | "root";
  children?: string[];
  body?: string;
  data?: T;
};
export type DNodeRawProps<T extends DNodeData = DNodeData> = Required<
  DNodeRawOpts<T>
>;

export type IDNodeOpts<T = DNodeData> = Omit<
  DNodeRawOpts<T>,
  "parent" | "children"
> & {
  type: IDNodeType;
  parent?: IDNode<T> | null;
  children?: IDNode<T>[];
};
export type IDNodeProps<T = DNodeData> = Required<IDNodeOpts<T>>;

export type IDNode<T = DNodeData> = IDNodeProps<T> & {
  // generated
  nodes: IDNode<T>[];
  // NOTE: currently equal to `fname`
  path: string;
  basename: string;
  logicalPath: string;
  // how to display path in lookup
  // used to special case `root` node to be empty string
  queryPath: string;
  // the root of domain
  domain: IDNode<T>;
  // absolute url to node id
  url: string;
  label: string;

  equal(node: IDNode<T>): boolean;
  // match(identifier: string): boolean;
  addChild(node: IDNode<T>): void;
  /**
   * Render body for rich-markdown-editor
   */
  renderBody(): string;
  toDocument(): any;
  toRawProps(): DNodeRawProps<T>;
  toRawPropsRecursive(): DNodeRawProps<T>[];
  validate(): boolean;
};
export type DNodeRawDict<T = DNodeData> = { [id: string]: DNodeRawProps<T> };
export type DNodeDict<T = DNodeData> = { [id: string]: IDNode<T> };

// --- Notes
export type NoteRawProps = DNodeRawProps<NoteData>;
export type INoteOpts = Omit<IDNodeOpts<NoteData>, "type">;
export type INoteProps = Required<INoteOpts>;
export type INote = INoteProps;
export type NoteDict = { [id: string]: Note };

// --- Schema
export type SchemaRawOpts = DNodeRawOpts<SchemaData> &
  Required<Pick<DNodeRawOpts<SchemaData>, "id" | "fname">>;
export type SchemaRawProps = DNodeRawProps<SchemaData>;
export type ISchemaOpts = Omit<IDNodeOpts<SchemaData>, "type">;
export type ISchemaProps = Required<ISchemaOpts>;
export type ISchema = ISchemaProps;
export type SchemaDict = { [id: string]: Schema };

// === Engine Types
export interface Resp<T> {
  data: T;
  error?: Error | null;
}

export type EngineGetResp<T = DNodeData> = Resp<IDNode<T>>;
export type EngineQueryResp<T = DNodeData> = Resp<IDNode<T>[]>;
export type StoreGetResp<T = DNodeData> = Resp<DNodeRawProps<T>>;
export type StoreQueryResp<T = DNodeData> = Resp<DNodeRawProps<T>[]>;

export interface Scope {
  username: string;
}

export interface NodeWriteOpts {
  /**
   * If newNode, will add it to the parent
   */
  newNode?: boolean;
  body?: string;
  /**
   * See QueryOpts.stub
   */
  stub?: boolean;
}

export interface QueryOpts {
  fullNode?: boolean;
  /**
   * Just get one result
   */
  queryOne?: boolean;
  /**
   * Use with `createIfNew`
   * If true, create a stub node
   */
  stub?: boolean;
  /**
   * If node does not exist, create it?
   */
  createIfNew?: boolean;
  // --- hints
  // DEPPRECATE
  webClient?: boolean;
  initialQuery?: boolean;
  mode?: QueryMode;
}

export interface DEngineParser<TOpts = any> {
  // parse: <T>(content: any, mode: QueryMode, opts: TOpts) => DNodeRawProps<T>[];
  parseSchema(data: any, opts: TOpts): SchemaRawProps[];
  parseNote(data: any, opts: TOpts): NoteRawProps[];
}

export type DEngineStoreWriteOpts = {
  /**
   * If set, don't write to file
   */
  stub?: boolean;
};

export interface DEngineStore<T = DNodeData, O = any> {
  opts: O;
  // fetchInitial: () => DNodeDict;
  delete: (id: string) => Promise<void>;
  get: (scope: Scope, id: string, opts?: QueryOpts) => Promise<StoreGetResp<T>>;
  query: (
    scope: Scope,
    queryString: string,
    mode: QueryMode,
    opts?: QueryOpts
  ) => Promise<EngineQueryResp<T>>;
  write: <T>(
    scope: Scope,
    node: IDNode<T>,
    opts?: DEngineStoreWriteOpts
  ) => Promise<void>;
}

export type DEngineMode = "exact" | "fuzzy";

export type DEngineOpts = {
  root: string;
  mode?: DEngineMode;
};

/**
 * Query: path based
 * Get: id based
 */
export interface DEngine {
  notes: NoteDict;
  schemas: SchemaDict;
  opts: Required<DEngineOpts>;

  /**
   * Load all nodes
   */
  init: () => Promise<void>;

  delete: (id: string) => Promise<void>;
  /**
   * Get node based on id
   * get(id: ...)
   */
  get: (
    scope: Scope,
    id: string,
    mode: QueryMode,
    opts?: QueryOpts
  ) => Promise<EngineGetResp<DNodeData>>;

  // getBatch: (scope: Scope, ids: string[]) => Promise<NodeGetBatchResp>;

  /**
   * Get node based on query
   * query(scope: {username: lukesernau}, queryString: "project", nodeType: note)
   * - []
   * - [Node(id: ..., title: project, children: [])]
   */
  query: (
    scope: Scope,
    queryString: string,
    mode: QueryMode,
    opts?: QueryOpts
  ) => Promise<EngineQueryResp<DNodeData>>;

  write: (
    scope: Scope,
    node: IDNode<DNodeData>,
    opts?: NodeWriteOpts
  ) => Promise<void>;
  // /**
  //  * Write node to db
  //  */
  // write: (scope: Scope, node: IDNode) => Promise<void>;

  // /**
  //  * Write list of nodes
  //  */
  // writeBatch: (scope: Scope, nodes: DNodeDict) => Promise<void>;
}
