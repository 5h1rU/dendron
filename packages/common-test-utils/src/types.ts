import { DEngineClientV2, DVault } from "@dendronhq/common-all";

export type TestResult = {
  actual: any;
  expected: any;
  msg?: string;
};

export type WorkspaceOpts = {
  wsRoot: string;
  vaults: DVault[];
};

export type SetupHookFunction<T = any> = (
  opts: {
    engine?: DEngineClientV2;
  } & WorkspaceOpts
) => Promise<T>;

export type PostSetupHookFunction<T = any> = (
  opts: {
    engine: DEngineClientV2;
  } & WorkspaceOpts
) => Promise<T>;
