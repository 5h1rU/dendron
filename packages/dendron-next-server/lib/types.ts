
// export type StageEnv = {
//   ENGINE_ENDPOINT_PORT?: number;
// };

export enum CONFIG_KEY {
  ENGINE_ENDPOINT_PORT = "ENGINE_ENDPOINT_PORT",
  REMOTE_API_ENDPOINT = "REMOTE_API_ENDPOINT",
}

export type StageEnv = typeof CONFIG_KEY;