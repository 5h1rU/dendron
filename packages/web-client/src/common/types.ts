export interface Node<TData> {
  /**
   * Local ID
   */
  id: string;
  data: TData;
  // currently string, in the future, this might be a tree of nodes
  body?: string;
  parent: NodeStub<TData> | null;
  children: NodeStub<TData>[];
}

/**
 */

type NodeStub<TData> = Omit<Node<TData>, "parent" | "children" | "body">;
export type NodeStubDict<TData> = { [id: string]: NodeStub<TData> };
export type NodeDict<TData> = { [id: string]: Node<TData> };

export function fromStubs<TData>(stubs: NodeStubDict<TData>): NodeDict<TData> {
  const keys = Object.keys(stubs).sort();
  const out: NodeDict<TData> = {};
  const helper = (i: number, parent: Node<TData>) => {
    while (i < keys.length && keys[i].startsWith(parent.id)) {
      parent.children.push({ ...stubs[keys[i]] });
      out[keys[i]] = {
        ...stubs[keys[i]],
        parent: { ...stubs[parent.id] },
        children: [],
      };
      i++;
      helper(i, out[keys[i]]);
    }
  };
  out[keys[0]] = { ...stubs[keys[0]], parent: null, children: [] };
  helper(1, out[keys[0]]);
  return out;
}

interface NodeData {
  title: string;
  desc: string;
}
export interface SchemaData extends NodeData {
  aliases?: string[];
  kind?: SchemaNodeKind;
  // match: SchemaNodeMatchRule[];
  choices?: { [key: string]: SchemaNodeStub };
  // links: SchemaNodeLink[];
  /**
   * If namespace exists, display it here
   */
  //namespace?: string;
  //   language:
  //     kind: namespace
  //     choices:
  //         python:
  //         ruby:
  //         ts:
  //     children:
  //         data
  //         flow
}

export interface NoteData extends NodeData {}
export type NoteNode = Node<NoteData>;
export type NoteNodeStub = NodeStub<NoteData>;
export type NoteNodeDict = { [key: string]: NoteNode };
export type NoteStubDict = { [key: string]: NoteNodeStub };

export type SchemaDataKey = keyof SchemaData;

export type RequiredSchemaDataKey = "title" | "desc";
export const RequiredSchemaDataKeyValues: RequiredSchemaDataKey[] = [
  "title",
  "desc",
];

export type OptionalSchemaDataKey = Exclude<
  keyof SchemaData,
  RequiredSchemaDataKey
>;
export type SchemaNodeDict = { [key: string]: SchemaNode };
export type SchemaNodeKind = "namespace";
export type SchemaNodeStub = NodeStub<SchemaData>;
export type SchemaNode = Node<SchemaData>;

export type SchemaYAMLRaw = {
  name: string;
  schema: { [key: string]: SchemaYAMLEntryRaw } | { root: SchemaYAMLEntryRaw };
};
export type SchemaYAMLEntryRaw = SchemaData & {
  children: { [key: string]: any };
};

/*
global:
  schema:
    output:
      aliases: [out]
      desc: anything I want to put out. formerly known as report
    usecase:
      alias: use
      desc: describes a way something is used
    gotcha:
      desc: something novel
    journal:
      desc: continuously updated
      kind: namespace
      choices:
        meet: 
        design:
        changelog:
    internal:
      desc: internal impletation
    templates:
      alias: temp
      desc: re-usable templates
    notes:
      desc: notes on something
    todo:
      desc: "things to update"
      children:
        tasks:
        questions:
    questions:
      desc: "questions i have"
    ref:
      desc: catchall
*/

type NodeType = "stub" | "full";
type DataType = "schema" | "note";

type NodeGetResp<T> = {
  item: Node<T> | NodeStub<T>;
  nodeType: NodeType;
  dataType: DataType;
};

export type NodeGetRootResp<T> = {
  item: Node<T>;
  dataType: DataType;
};

interface NodeQueryResp<T> {
  items: NodeDict<T>;
  nodeType: NodeType;
  dataType: DataType;
  error: Error;
}
interface Scope {
  username: string;
}

export interface DendronEngine {
  /**
   * Get node based on id
   * get(id: ...)
   */
  get: <T>(
    scope: Scope,
    id: string,
    nodeType: NodeType,
    dataType: DataType
  ) => Promise<NodeGetResp<T>>;

  // load alphacortex
  /**
   * NodeStorageAPI.getRoot()
   * {
   *   parent: null
   *   children:
   *     - Alphacortex
   * }
   *
   * welcome page
   * sider
   */
  // getRoot: <T>(scope: Scope, dataType: DataType) => Promise<NodeGetRootResp<T>>;

  /**
   * Get node based on query
   * query(scope: {username: lukesernau}, queryString: "Project", nodeType: stub, dataType: note})
   * - []
   * - [Node(id: ..., title: project, children: [])]
   *
   */
  query: <T>(
    scope: Scope,
    queryString: string,
    nodeType: NodeType,
    dataType: DataType
  ) => Promise<NodeQueryResp<T>>;

  /**
   * Write node to db
   */
  write: <T>(scope: Scope, node: Node<T>, dataType: DataType) => Promise<void>;

  /**
   * Write list of nodes
   */
  writeBatch: <T>(
    scope: Scope,
    nodes: NodeDict<T>[],
    dataType: DataType
  ) => Promise<void>;
}

// convert to dictionary, in return
