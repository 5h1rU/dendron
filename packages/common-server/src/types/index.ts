import { NoteProps } from "@dendronhq/common-all";
import execa from "execa";
import _ from "lodash";

export type DHookFunction = (opts: {
  note: NoteProps;
  execa: typeof execa;
  _: typeof _;
}) => Promise<NoteProps>;
