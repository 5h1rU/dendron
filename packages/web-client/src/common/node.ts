import {
  SchemaDataKey,
  SchemaNode,
  SchemaNodeStub,
  SchemaYAML,
  SchemaYAMLEntry,
} from "./types";

import YAML from "yamljs";
import _ from "lodash";

interface DataDefaults {
  default: any;
}

export class SchemaStubWrapper {
  static fromSchemaNode(node: SchemaNode): SchemaNodeStub {
    return _.omit(node, "children", "parent");
  }

  static fromSchemaYAMLEntry(entry: SchemaYAMLEntry): SchemaNodeStub {
    const { id } = entry;
    const schemaDataKeysDefaults: {
      [key in SchemaDataKey]: any;
    } = {
      aliases: [],
      kind: undefined,
      choices: [],
      title: id,
      desc: "",
    };
    //const title = entry.title ? entry.title : entry.id;

    const data = _.defaults(
      {},
      _.omit(entry, "children"),
      schemaDataKeysDefaults
    );
    const logicalId = id;
    const schemaNode = { data, id, logicalId };
    return schemaNode;
  }
}

export class SchemaNodeWrapper {
  static fromSchemaYAMLEntry(entry: SchemaYAMLEntry): SchemaNode {
    const schemaStub = SchemaStubWrapper.fromSchemaYAMLEntry(entry);
    const parent = null;
    const children: any[] = [];
    const schemaNode = { ...schemaStub, parent, children };
    return schemaNode;
  }

  // static deserialize(yamlString: string): SchemaTree {
  //   const schema: SchemaYAML = YAML.parse(yamlString);
  //   const tree = SchemaTree.fromSchemaYAML(schema);
  //   return tree;
  // }
}

export class SchemaTree {
  public name: string;
  public root: SchemaNode;
  public nodes: { [key: string]: SchemaNode };

  constructor(name: string, root: SchemaNode) {
    this.name = name;
    this.root = root;
    this.nodes = {};
  }

  addChild(child: SchemaNode, parent: SchemaNode) {
    const parentNode = this.nodes[parent.logicalId];
    if (_.isUndefined(parentNode)) {
      throw `no parent with ${parent.id} found`;
    }
    const childStub = SchemaStubWrapper.fromSchemaNode(child);
    parentNode.children.push(childStub);
    this.nodes[child.logicalId] = child;
  }

  static fromSchemaYAML(yamlString: string): SchemaTree {
    const schemaYAML: SchemaYAML = YAML.parse(yamlString);
    const { name, schema } = schemaYAML;

    const root = SchemaNodeWrapper.fromSchemaYAMLEntry({
      ...schema.root,
      id: "root",
    });
    const tree = new SchemaTree(name, root);
    let parent = root;
    _.map(tree.root.children, (v: any, childId: string) => {
      // @ts-ignore
      const entry: SchemaYAMLEntry = schema[childId];

      const childNode = SchemaNodeWrapper.fromSchemaYAMLEntry({
        ...entry,
        id: entry.id,
      });
      tree.addChild(childNode, parent);
    });
    // tree.root.children.forEach((childId: string) => {
    //   const childEntry = schema[childId];
    // });
    return tree;
  }

  toD3Tree() {
    // TODO
    return;
  }
}
