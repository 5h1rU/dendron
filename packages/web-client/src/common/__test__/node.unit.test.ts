import { SchemaNode } from "../types";
import { SchemaTree } from "../node";

const rootSchemaNode: SchemaNode = {
  id: "root",
  logicalId: "root",
  children: [],
  parent: null,
  data: { title: "root", desc: "root" },
};

const fooSchemaNode: SchemaNode = {
  id: "foo",
  logicalId: "foo",
  children: [],
  parent: null,
  data: { title: "foo", desc: "foo" },
};

const SAMPLE_YAML = `
  name: project
  schema:
      root:
        children:
          quickstart: 
          topic: 
          version: 
          dev: 
          features:
          rel:
      quickstart:
        desc: get started with project
      features:
        desc: what does it do
      ref:
        kind: namespace
        choices:
            competitors: 
            shortcuts:
      rel:
        desc: relative
      version:
        children:
          version-major: 
          version-minor: 
          version-breaking: 
      plan:
        children:
          requirements:
            alias: req
          timeline:
            desc: "how long will it take"
      version-major:
        desc: the major version
`;

describe("SchemaTree", () => {
  test("root", () => {
    const tree = new SchemaTree("root", rootSchemaNode);
    expect(tree).toMatchSnapshot("root-snap");
  });
  test("foo", () => {
    const tree = new SchemaTree("foo", fooSchemaNode);
    expect(tree).toMatchSnapshot("foo-snap");
  });
  test("root/foo", () => {
    const tree1 = new SchemaTree("root", rootSchemaNode);
    const tree2 = new SchemaTree("foo", fooSchemaNode);
    tree1.addSubTree(tree2, "root");
    expect(tree1).toMatchSnapshot("root/foo-snap");
  });
  test("fromSchemaYAML", () => {
    const tree = SchemaTree.fromSchemaYAML(SAMPLE_YAML);
    expect(tree).toMatchSnapshot("fromSchemaYAML");
  });
});
