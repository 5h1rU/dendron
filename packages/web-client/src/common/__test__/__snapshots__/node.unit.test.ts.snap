// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchemaTree foo: foo-snap 1`] = `
SchemaTree {
  "name": "foo",
  "nodes": Object {
    "foo": Object {
      "children": Array [],
      "data": Object {
        "desc": "foo",
        "title": "foo",
      },
      "id": "foo",
      "logicalId": "foo",
      "parent": null,
    },
  },
  "root": Object {
    "children": Array [],
    "data": Object {
      "desc": "foo",
      "title": "foo",
    },
    "id": "foo",
    "logicalId": "foo",
    "parent": null,
  },
}
`;

exports[`SchemaTree fromSchemaYAML: fromSchemaYAML 1`] = `
SchemaTree {
  "name": "project",
  "nodes": Object {
    "dev": Object {
      "children": Array [],
      "data": Object {
        "aliases": Array [],
        "choices": Array [],
        "desc": "",
        "kind": undefined,
        "title": "dev",
      },
      "id": "dev",
      "logicalId": "dev",
      "parent": null,
    },
    "features": Object {
      "children": Array [],
      "data": Object {
        "aliases": Array [],
        "choices": Array [],
        "desc": "what does it do",
        "kind": undefined,
        "title": "features",
      },
      "id": "features",
      "logicalId": "features",
      "parent": null,
    },
    "project": Object {
      "children": Array [
        Object {
          "data": Object {
            "aliases": Array [],
            "choices": Array [],
            "desc": "",
            "kind": undefined,
            "title": "quickstart",
          },
          "id": "quickstart",
          "logicalId": "quickstart",
        },
        Object {
          "data": Object {
            "aliases": Array [],
            "choices": Array [],
            "desc": "",
            "kind": undefined,
            "title": "topic",
          },
          "id": "topic",
          "logicalId": "topic",
        },
        Object {
          "data": Object {
            "aliases": Array [],
            "choices": Array [],
            "desc": "",
            "kind": undefined,
            "title": "version",
          },
          "id": "version",
          "logicalId": "version",
        },
        Object {
          "data": Object {
            "aliases": Array [],
            "choices": Array [],
            "desc": "",
            "kind": undefined,
            "title": "dev",
          },
          "id": "dev",
          "logicalId": "dev",
        },
        Object {
          "data": Object {
            "aliases": Array [],
            "choices": Array [],
            "desc": "",
            "kind": undefined,
            "title": "features",
          },
          "id": "features",
          "logicalId": "features",
        },
        Object {
          "data": Object {
            "aliases": Array [],
            "choices": Array [],
            "desc": "",
            "kind": undefined,
            "title": "rel",
          },
          "id": "rel",
          "logicalId": "rel",
        },
      ],
      "data": Object {
        "aliases": Array [],
        "choices": Array [],
        "desc": "",
        "kind": undefined,
        "title": "project",
      },
      "id": "project",
      "logicalId": "project",
      "parent": null,
    },
    "quickstart": Object {
      "children": Array [],
      "data": Object {
        "aliases": Array [],
        "choices": Array [],
        "desc": "get started with project",
        "kind": undefined,
        "title": "quickstart",
      },
      "id": "quickstart",
      "logicalId": "quickstart",
      "parent": null,
    },
    "rel": Object {
      "children": Array [],
      "data": Object {
        "aliases": Array [],
        "choices": Array [],
        "desc": "relative",
        "kind": undefined,
        "title": "rel",
      },
      "id": "rel",
      "logicalId": "rel",
      "parent": null,
    },
    "topic": Object {
      "children": Array [],
      "data": Object {
        "aliases": Array [],
        "choices": Array [],
        "desc": "",
        "kind": undefined,
        "title": "topic",
      },
      "id": "topic",
      "logicalId": "topic",
      "parent": null,
    },
    "version": Object {
      "children": Array [
        Object {
          "data": Object {
            "aliases": Array [],
            "choices": Array [],
            "desc": "",
            "kind": undefined,
            "title": "version-major",
          },
          "id": "version-major",
          "logicalId": "version-major",
        },
        Object {
          "data": Object {
            "aliases": Array [],
            "choices": Array [],
            "desc": "",
            "kind": undefined,
            "title": "version-minor",
          },
          "id": "version-minor",
          "logicalId": "version-minor",
        },
        Object {
          "data": Object {
            "aliases": Array [],
            "choices": Array [],
            "desc": "",
            "kind": undefined,
            "title": "version-breaking",
          },
          "id": "version-breaking",
          "logicalId": "version-breaking",
        },
      ],
      "data": Object {
        "aliases": Array [],
        "choices": Array [],
        "desc": "",
        "kind": undefined,
        "title": "version",
      },
      "id": "version",
      "logicalId": "version",
      "parent": null,
    },
    "version-breaking": Object {
      "children": Array [],
      "data": Object {
        "aliases": Array [],
        "choices": Array [],
        "desc": "",
        "kind": undefined,
        "title": "version-breaking",
      },
      "id": "version-breaking",
      "logicalId": "version-breaking",
      "parent": null,
    },
    "version-major": Object {
      "children": Array [],
      "data": Object {
        "aliases": Array [],
        "choices": Array [],
        "desc": "the major version",
        "kind": undefined,
        "title": "version-major",
      },
      "id": "version-major",
      "logicalId": "version-major",
      "parent": null,
    },
    "version-minor": Object {
      "children": Array [],
      "data": Object {
        "aliases": Array [],
        "choices": Array [],
        "desc": "",
        "kind": undefined,
        "title": "version-minor",
      },
      "id": "version-minor",
      "logicalId": "version-minor",
      "parent": null,
    },
  },
  "root": Object {
    "children": Array [
      Object {
        "data": Object {
          "aliases": Array [],
          "choices": Array [],
          "desc": "",
          "kind": undefined,
          "title": "quickstart",
        },
        "id": "quickstart",
        "logicalId": "quickstart",
      },
      Object {
        "data": Object {
          "aliases": Array [],
          "choices": Array [],
          "desc": "",
          "kind": undefined,
          "title": "topic",
        },
        "id": "topic",
        "logicalId": "topic",
      },
      Object {
        "data": Object {
          "aliases": Array [],
          "choices": Array [],
          "desc": "",
          "kind": undefined,
          "title": "version",
        },
        "id": "version",
        "logicalId": "version",
      },
      Object {
        "data": Object {
          "aliases": Array [],
          "choices": Array [],
          "desc": "",
          "kind": undefined,
          "title": "dev",
        },
        "id": "dev",
        "logicalId": "dev",
      },
      Object {
        "data": Object {
          "aliases": Array [],
          "choices": Array [],
          "desc": "",
          "kind": undefined,
          "title": "features",
        },
        "id": "features",
        "logicalId": "features",
      },
      Object {
        "data": Object {
          "aliases": Array [],
          "choices": Array [],
          "desc": "",
          "kind": undefined,
          "title": "rel",
        },
        "id": "rel",
        "logicalId": "rel",
      },
    ],
    "data": Object {
      "aliases": Array [],
      "choices": Array [],
      "desc": "",
      "kind": undefined,
      "title": "project",
    },
    "id": "project",
    "logicalId": "project",
    "parent": null,
  },
}
`;

exports[`SchemaTree root/foo: root/foo-snap 1`] = `
SchemaTree {
  "name": "root",
  "nodes": Object {
    "foo": Object {
      "children": Array [],
      "data": Object {
        "desc": "foo",
        "title": "foo",
      },
      "id": "foo",
      "logicalId": "foo",
      "parent": null,
    },
    "root": Object {
      "children": Array [
        Object {
          "data": Object {
            "desc": "foo",
            "title": "foo",
          },
          "id": "foo",
          "logicalId": "foo",
        },
      ],
      "data": Object {
        "desc": "root",
        "title": "root",
      },
      "id": "root",
      "logicalId": "root",
      "parent": null,
    },
  },
  "root": Object {
    "children": Array [
      Object {
        "data": Object {
          "desc": "foo",
          "title": "foo",
        },
        "id": "foo",
        "logicalId": "foo",
      },
    ],
    "data": Object {
      "desc": "root",
      "title": "root",
    },
    "id": "root",
    "logicalId": "root",
    "parent": null,
  },
}
`;

exports[`SchemaTree root: root-snap 1`] = `
SchemaTree {
  "name": "root",
  "nodes": Object {
    "root": Object {
      "children": Array [],
      "data": Object {
        "desc": "root",
        "title": "root",
      },
      "id": "root",
      "logicalId": "root",
      "parent": null,
    },
  },
  "root": Object {
    "children": Array [],
    "data": Object {
      "desc": "root",
      "title": "root",
    },
    "id": "root",
    "logicalId": "root",
    "parent": null,
  },
}
`;
